File: python/mirascope/llm/retries/retry_stream_responses.py
"""Retry stream response wrappers that add retry capabilities to streaming LLM responses."""

from collections.abc import AsyncIterator, Awaitable, Callable, Iterator
from typing import TYPE_CHECKING, overload

from ..content import (
    AssistantContentChunk,
)
from ..exceptions import StreamRestarted
from ..formatting import FormattableT
from ..messages import UserContent
from ..models import Model
from ..responses import (
    AsyncStreamResponse,
    StreamResponse,
)
from .utils import RetryFailure

if TYPE_CHECKING:
    from .retry_models import RetryModel


class RetryStreamResponse(StreamResponse[FormattableT]):
    """A streaming response wrapper that includes retry capabilities.

    Extends `StreamResponse` directly, copying all attributes from a wrapped response
    and adding retry configuration.

    This wraps a `StreamResponse` and adds automatic retry behavior when
    retryable errors occur during iteration. When a retry happens, a
    `StreamRestarted` exception is raised so the user can handle the restart
    (e.g., clear terminal output) before re-iterating.

    Note: Streaming does not currently support fallback models. Retries will
    only be attempted with the active model.

    Example:
        ```python
        response = retry_model.stream("Tell me a story")

        while True:
            try:
                for chunk in response.text_stream():
                    print(chunk, end="", flush=True)
                break  # Success
            except llm.StreamRestarted as e:
                print(f"\\n[Retry {e.attempt} after: {e.error}]")
        ```
    """

    _retry_model: "RetryModel"
    """The RetryModel providing retry configuration."""

    retry_failures: list[RetryFailure]
    """Failed attempts before success (empty if first attempt succeeded)."""

    _stream_fn: Callable[[Model], StreamResponse[FormattableT]]

    def __init__(
        self,
        retry_model: "RetryModel",
        stream_fn: Callable[[Model], StreamResponse[FormattableT]],
    ) -> None:
        """Initialize a RetryStreamResponse.

        Args:
            retry_model: The RetryModel providing retry configuration.
            stream_fn: Function that creates a stream from a Model.
        """
        # Create the initial stream using the active model and copy all attributes
        active_model = retry_model.get_active_model()
        initial_stream = stream_fn(active_model)
        for key, value in initial_stream.__dict__.items():
            object.__setattr__(self, key, value)

        self._retry_model = retry_model
        self._stream_fn = stream_fn
        self.retry_failures = []

    @property
    def model(self) -> "RetryModel":
        """A RetryModel with parameters matching this response."""
        return self._retry_model

    def _reset_stream(self) -> None:
        """Reset to a fresh stream for a new retry attempt."""
        active_model = self._retry_model.get_active_model()
        new_stream = self._stream_fn(active_model)
        # Copy all attributes from the new stream
        for key, value in new_stream.__dict__.items():
            object.__setattr__(self, key, value)

    def chunk_stream(self) -> Iterator[AssistantContentChunk]:
        """Returns an iterator that yields content chunks with retry support.

        If a retryable error occurs during iteration, the stream is reset
        and a `StreamRestarted` exception is raised. The user should catch
        this exception and re-iterate to continue from the new attempt.

        Raises:
            StreamRestarted: When the stream is reset for a retry attempt.
            Exception: The underlying error if max retries are exhausted.
        """
        config = self._retry_model.retry_config
        active_model = self._retry_model.get_active_model()
        try:
            yield from super().chunk_stream()
            # Success - no need to record anything
        except config.retry_on as e:
            self.retry_failures.append(RetryFailure(model=active_model, exception=e))
            failure_count = len(self.retry_failures)

            if failure_count > config.max_retries:
                raise

            self._reset_stream()

            raise StreamRestarted(
                attempt=failure_count,
                error=e,
            ) from e

    @overload
    def resume(
        self: "RetryStreamResponse[None]", content: UserContent
    ) -> "RetryStreamResponse[None]": ...

    @overload
    def resume(
        self: "RetryStreamResponse[FormattableT]", content: UserContent
    ) -> "RetryStreamResponse[FormattableT]": ...

    def resume(
        self, content: UserContent
    ) -> "RetryStreamResponse[None] | RetryStreamResponse[FormattableT]":
        """Generate a new RetryStreamResponse using this response's messages with additional user content."""
        return self.model.resume_stream(response=self, content=content)


class AsyncRetryStreamResponse(AsyncStreamResponse[FormattableT]):
    """An async streaming response wrapper that includes retry capabilities.

    Extends `AsyncStreamResponse` directly, copying all attributes from a wrapped response
    and adding retry configuration.

    This wraps an `AsyncStreamResponse` and adds automatic retry behavior when
    retryable errors occur during iteration. When a retry happens, a
    `StreamRestarted` exception is raised so the user can handle the restart
    (e.g., clear terminal output) before re-iterating.

    Note: Streaming does not currently support fallback models. Retries will
    only be attempted with the active model.

    Example:
        ```python
        response = await retry_model.stream_async("Tell me a story")

        while True:
            try:
                async for chunk in response.text_stream():
                    print(chunk, end="", flush=True)
                break  # Success
            except llm.StreamRestarted as e:
                print(f"\\n[Retry {e.attempt} after: {e.error}]")
        ```
    """

    _retry_model: "RetryModel"
    """The RetryModel providing retry configuration."""

    retry_failures: list[RetryFailure]
    """Failed attempts before success (empty if first attempt succeeded)."""

    _stream_fn: Callable[[Model], Awaitable[AsyncStreamResponse[FormattableT]]]

    def __init__(
        self,
        retry_model: "RetryModel",
        stream_fn: Callable[[Model], Awaitable[AsyncStreamResponse[FormattableT]]],
        initial_stream: AsyncStreamResponse[FormattableT],
    ) -> None:
        """Initialize an AsyncRetryStreamResponse.

        Args:
            retry_model: The RetryModel providing retry configuration.
            stream_fn: Async function that creates a stream from a Model.
            initial_stream: The pre-awaited initial stream.
        """
        # Copy all attributes from the initial stream
        for key, value in initial_stream.__dict__.items():
            object.__setattr__(self, key, value)

        self._retry_model = retry_model
        self._stream_fn = stream_fn
        self.retry_failures = []

    @property
    def model(self) -> "RetryModel":
        """A RetryModel with parameters matching this response."""
        return self._retry_model

    async def _reset_stream(self) -> None:
        """Reset to a fresh stream for a new retry attempt."""
        active_model = self._retry_model.get_active_model()
        new_stream = await self._stream_fn(active_model)
        # Copy all attributes from the new stream
        for key, value in new_stream.__dict__.items():
            object.__setattr__(self, key, value)

    async def chunk_stream(self) -> AsyncIterator[AssistantContentChunk]:
        """Returns an async iterator that yields content chunks with retry support.

        If a retryable error occurs during iteration, the stream is reset
        and a `StreamRestarted` exception is raised. The user should catch
        this exception and re-iterate to continue from the new attempt.

        Raises:
            StreamRestarted: When the stream is reset for a retry attempt.
            Exception: The underlying error if max retries are exhausted.
        """
        config = self._retry_model.retry_config
        active_model = self._retry_model.get_active_model()
        try:
            async for chunk in super().chunk_stream():
                yield chunk
            # Success - no need to record anything
        except config.retry_on as e:
            self.retry_failures.append(RetryFailure(model=active_model, exception=e))
            failure_count = len(self.retry_failures)

            if failure_count > config.max_retries:
                raise

            await self._reset_stream()

            raise StreamRestarted(
                attempt=failure_count,
                error=e,
            ) from e

    @overload
    async def resume(
        self: "AsyncRetryStreamResponse[None]", content: UserContent
    ) -> "AsyncRetryStreamResponse[None]": ...

    @overload
    async def resume(
        self: "AsyncRetryStreamResponse[FormattableT]", content: UserContent
    ) -> "AsyncRetryStreamResponse[FormattableT]": ...

    async def resume(
        self, content: UserContent
    ) -> "AsyncRetryStreamResponse[None] | AsyncRetryStreamResponse[FormattableT]":
        """Generate a new AsyncRetryStreamResponse using this response's messages with additional user content."""
        return await self.model.resume_stream_async(response=self, content=content)

File: typescript/src/llm/retries/retry-stream-responses.ts
/**
 * Retry stream response wrappers that add retry capabilities to streaming LLM responses.
 */

import type { AssistantContentChunk } from "@/llm/content";
import type { ToolOutput } from "@/llm/content/tool-output";
import type { Context } from "@/llm/context";
import type { UserContent } from "@/llm/messages";
import type { Model } from "@/llm/models";
import type { Jsonable } from "@/llm/types/jsonable";

import { RetriesExhausted, StreamRestarted } from "@/llm/exceptions";
import { ContextStreamResponse } from "@/llm/responses/context-stream-response";
import { StreamResponse } from "@/llm/responses/stream-response";

import type { RetryConfig } from "./retry-config";
import type { RetryModel } from "./retry-model";

import { getRetryModelFromContext } from "./retry-model";
import { isRetryableError, type RetryFailure } from "./utils";

/**
 * Create a dummy async iterator for initial construction.
 * The actual iteration will be handled by the wrapped stream.
 */
async function* emptyIterator(): AsyncGenerator<never> {
  // Never yields anything - we override chunkStream to use the wrapped stream
}

/**
 * A streaming response wrapper that includes retry capabilities.
 *
 * This class wraps a `StreamResponse` and adds automatic retry behavior when
 * retryable errors occur during iteration. When a retry happens, a
 * `StreamRestarted` exception is raised so the user can handle the restart
 * (e.g., clear terminal output) before re-iterating.
 *
 * @template F - The type of the formatted output when using structured outputs.
 *
 * @example
 * ```typescript
 * const response = await retryModel.stream("Tell me a story");
 *
 * while (true) {
 *   try {
 *     for await (const text of response.textStream()) {
 *       process.stdout.write(text);
 *     }
 *     break; // Success
 *   } catch (e) {
 *     if (e instanceof llm.StreamRestarted) {
 *       console.log(`\nStream restarted: ${e.message}`);
 *     } else {
 *       throw e;
 *     }
 *   }
 * }
 * ```
 */
export class RetryStreamResponse<F = unknown> extends StreamResponse<F> {
  /**
   * The current RetryModel variant being used.
   */
  private _currentVariant: RetryModel;

  /**
   * Async iterator over model variants with backoff delays.
   */
  private _variantsIter: AsyncGenerator<RetryModel>;

  /**
   * Failed attempts before success (empty if first attempt succeeded).
   */
  readonly retryFailures: RetryFailure[] = [];

  /**
   * Function that creates a stream from a Model.
   */
  private readonly _streamFn: (model: Model) => Promise<StreamResponse<F>>;

  /**
   * The current underlying stream response.
   */
  private _currentStream: StreamResponse<F>;

  /**
   * Initialize a RetryStreamResponse.
   *
   * @param streamFn - Async function that creates a stream from a Model.
   * @param initialStream - The pre-awaited initial stream.
   * @param initialVariant - The first variant from the iterator.
   * @param variantsIter - The async iterator for remaining variants.
   */
  constructor(
    streamFn: (model: Model) => Promise<StreamResponse<F>>,
    initialStream: StreamResponse<F>,
    initialVariant: RetryModel,
    variantsIter: AsyncGenerator<RetryModel>,
  ) {
    // Initialize with the initial stream's properties
    super({
      providerId: initialStream.providerId,
      modelId: initialStream.modelId,
      providerModelName: initialStream.providerModelName,
      params: initialStream.params,
      tools: initialStream.toolkit,
      format: initialStream.format,
      inputMessages: initialStream.inputMessages,
      chunkIterator: emptyIterator(),
    });

    this._currentVariant = initialVariant;
    this._variantsIter = variantsIter;
    this._streamFn = streamFn;
    this._currentStream = initialStream;
  }

  /**
   * A RetryModel with parameters matching this response.
   *
   * If a model is set in context (via `withModel()`), that model is used
   * instead, wrapped with this response's retry config.
   */
  override get model(): Promise<RetryModel> {
    return Promise.resolve(getRetryModelFromContext(this._currentVariant));
  }

  /**
   * The retry configuration for this response.
   */
  get retryConfig(): RetryConfig {
    return this._currentVariant.retryConfig;
  }

  /**
   * Reset to a fresh stream for a new retry attempt.
   */
  private async _resetStream(variant: RetryModel): Promise<void> {
    this._currentStream = await this._streamFn(variant.getActiveModel());
  }

  /**
   * Stream content chunks with retry support.
   *
   * If a retryable error occurs during iteration, the stream is reset
   * and a `StreamRestarted` exception is raised. The user should catch
   * this exception and re-iterate to continue from the new attempt.
   *
   * @throws StreamRestarted - When the stream is reset for a retry attempt.
   * @throws RetriesExhausted - If all retry attempts fail.
   */
  override async *chunkStream(): AsyncGenerator<AssistantContentChunk> {
    try {
      for await (const chunk of this._currentStream.chunkStream()) {
        yield chunk;
      }
    } catch (error) {
      /* v8 ignore next 2 -- defensive: errors are always Error instances */
      if (!(error instanceof Error)) throw error;
      if (!isRetryableError(error, this.retryConfig)) throw error;

      const failure: RetryFailure = {
        model: this._currentVariant.getActiveModel(),
        exception: error,
      };
      this.retryFailures.push(failure);

      // Try to get next variant (handles backoff and fallback)
      const next = await this._variantsIter.next();
      if (next.done) {
        throw new RetriesExhausted(this.retryFailures);
      }

      this._currentVariant = next.value;
      await this._resetStream(this._currentVariant);

      throw new StreamRestarted(failure);
    }
  }

  /**
   * Generate a new RetryStreamResponse using this response's messages with additional user content.
   *
   * @param content - The new user message content to append to the message history.
   * @returns A new RetryStreamResponse for streaming the extended conversation.
   */
  override async resume(content: UserContent): Promise<RetryStreamResponse<F>> {
    const model = await this.model;
    return model.resumeStream(this, content) as Promise<RetryStreamResponse<F>>;
  }

  /**
   * Execute all tool calls in this response using the registered tools.
   *
   * Note: The stream must be consumed before calling executeTools() to ensure
   * all tool calls have been received.
   *
   * @returns An array of ToolOutput objects, one for each tool call.
   */
  override async executeTools(): Promise<ToolOutput<Jsonable>[]> {
    return Promise.all(
      this.toolCalls.map((call) => this.toolkit.execute(call)),
    );
  }
}

/**
 * A context-aware streaming response wrapper that includes retry capabilities.
 *
 * This class wraps a `ContextStreamResponse` and adds automatic retry behavior when
 * retryable errors occur during iteration.
 *
 * @template DepsT - The type of dependencies in the context.
 * @template F - The type of the formatted output when using structured outputs.
 */
export class ContextRetryStreamResponse<
  DepsT = unknown,
  F = unknown,
> extends ContextStreamResponse<DepsT, F> {
  /**
   * The current RetryModel variant being used.
   */
  private _currentVariant: RetryModel;

  /**
   * Async iterator over model variants with backoff delays.
   */
  private _variantsIter: AsyncGenerator<RetryModel>;

  /**
   * Failed attempts before success (empty if first attempt succeeded).
   */
  readonly retryFailures: RetryFailure[] = [];

  /**
   * Function that creates a stream from a Model.
   */
  private readonly _streamFn: (
    model: Model,
  ) => Promise<ContextStreamResponse<DepsT, F>>;

  /**
   * The current underlying stream response.
   */
  private _currentStream: ContextStreamResponse<DepsT, F>;

  /**
   * Initialize a ContextRetryStreamResponse.
   *
   * @param streamFn - Async function that creates a stream from a Model.
   * @param initialStream - The pre-awaited initial stream.
   * @param initialVariant - The first variant from the iterator.
   * @param variantsIter - The async iterator for remaining variants.
   */
  constructor(
    streamFn: (model: Model) => Promise<ContextStreamResponse<DepsT, F>>,
    initialStream: ContextStreamResponse<DepsT, F>,
    initialVariant: RetryModel,
    variantsIter: AsyncGenerator<RetryModel>,
  ) {
    // Initialize with the initial stream's properties
    super({
      providerId: initialStream.providerId,
      modelId: initialStream.modelId,
      providerModelName: initialStream.providerModelName,
      params: initialStream.params,
      tools: initialStream.contextToolkit,
      format: initialStream.format,
      inputMessages: initialStream.inputMessages,
      chunkIterator: emptyIterator(),
    });

    this._currentVariant = initialVariant;
    this._variantsIter = variantsIter;
    this._streamFn = streamFn;
    this._currentStream = initialStream;
  }

  /**
   * A RetryModel with parameters matching this response.
   */
  override get model(): Promise<RetryModel> {
    return Promise.resolve(getRetryModelFromContext(this._currentVariant));
  }

  /**
   * The retry configuration for this response.
   */
  get retryConfig(): RetryConfig {
    return this._currentVariant.retryConfig;
  }

  /**
   * Reset to a fresh stream for a new retry attempt.
   */
  private async _resetStream(variant: RetryModel): Promise<void> {
    this._currentStream = await this._streamFn(variant.getActiveModel());
  }

  /**
   * Stream content chunks with retry support.
   */
  override async *chunkStream(): AsyncGenerator<AssistantContentChunk> {
    try {
      for await (const chunk of this._currentStream.chunkStream()) {
        yield chunk;
      }
    } catch (error) {
      /* v8 ignore next 2 -- defensive: errors are always Error instances */
      if (!(error instanceof Error)) throw error;
      if (!isRetryableError(error, this.retryConfig)) throw error;

      const failure: RetryFailure = {
        model: this._currentVariant.getActiveModel(),
        exception: error,
      };
      this.retryFailures.push(failure);

      // Try to get next variant (handles backoff and fallback)
      const next = await this._variantsIter.next();
      if (next.done) {
        throw new RetriesExhausted(this.retryFailures);
      }

      this._currentVariant = next.value;
      await this._resetStream(this._currentVariant);

      throw new StreamRestarted(failure);
    }
  }

  /**
   * Generate a new ContextRetryStreamResponse using this response's messages with additional user content.
   *
   * @param ctx - A Context with the required deps type.
   * @param content - The new user message content to append to the message history.
   * @returns A new ContextRetryStreamResponse for streaming the extended conversation.
   */
  override async resume(
    ctx: Context<DepsT>,
    content: UserContent,
  ): Promise<ContextRetryStreamResponse<DepsT, F>> {
    const model = await this.model;
    return model.contextResumeStream(ctx, this, content) as Promise<
      ContextRetryStreamResponse<DepsT, F>
    >;
  }

  /**
   * Execute all tool calls in this response using the registered context tools.
   *
   * @param ctx - The context containing dependencies to pass to tools.
   * @returns An array of ToolOutput objects, one for each tool call.
   */
  override async executeTools(
    ctx: Context<DepsT>,
  ): Promise<ToolOutput<Jsonable>[]> {
    return this.contextToolkit.executeAll(ctx, this.toolCalls);
  }
}
